fix(wstring): Correct out-of-range indexing and is_char_boundary

flub opened the pull request:
Fixes a bug when detecting UTF-16 surrogate pairs. Instead of checking
the full 6 bits, only four bits were checked. This leads to panics when
slicing valid wstrings at character boundaries.

Additionally, when indexing beyond the size of the slice, we should
error clearly. This is done by the `is_char_boundary()` check, which is
supposed to handle character boundaries outside of the slice len
gracefully by returning false instead of panicking. Fixing this fixes
all the indexing.

Commit:
fix: Correct out-of-range indexing and is_char_boundary

When indexing beyond the size of the slice, we should error clearly.
This is done by the is_char_boundary() check, which is supposed to
handle character boundaries outside of the slice len gracefully by
returning false instead of panicking.  Fixing this fixes all the
indexing.

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -121,7 +121,7 @@ impl WStr {
             return true;
         }
-        if index % 2 != 0 {
+        if index % 2 != 0 || index > self.len() {
```
If you want panic-less code below, you can also do this:

```rust
self.raw.get(index..index+2).map_or(false, |s| !is_trailing_surrogate(u16::copy_from_le_bytes(s))
```

Unfortunately, that does not get rid of the panicking in `copy_from_le_bytes`, but at least from the slice access. It would be phenomenal, if we could get rid of the `assert!` in `copy_from_le_bytes` somehow, but I think this is only possible if you change the signature to return `Option<u16>`:

```rust
self.raw
  .get(index..)
  .and_then(u16::copy_from_le_bytes)
  .map_or(false, |u| !is_trailing_surrogate(u))
```

Overall, I think this would be the cleanest change.

Commit:
Avoid some more panicking code

This makes the u16::copy_from_le_bytes() return an option which allows
us to write a little more code without panics without much if any
performance tradeoff.  The readability suffers though.

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -279,7 +282,7 @@ impl<'a> Iterator for WStrChars<'a> {
         // No need to fully construct all characters
         self.chunks
-            .filter(|bb| !is_trailing_surrogate(u16::copy_from_le_bytes(bb)))
+            .filter(|bb| !is_trailing_surrogate(u16::copy_from_le_bytes(bb).unwrap()))
```
Would this be better?

```suggestion
            .filter_map(|chunk| u16::copy_from_le_bytes(chunk))
            .filter(|unit| !is_trailing_surrogate(unit))
```

flub replied:
heh, i did consider that.  i just feels so ugly and confusing because you're not actually filter_mapping, you're just unpacking an option which you know can be unpacked.  Same with the other places of this commit that add `?`, in a way `.unwrap()` would have been clearer there.

jan-auer replied:
I mean, it's up to you. If you know that it can't go wrong, I think it's fair to use an infallible API.

jan-auer committed:
fix(wstring): Detect leading surrogates correctly

jan-auer commented on relay-wstring/src/lib.rs:
```diff
@@ -380,10 +408,10 @@ impl fmt::Display for WStr {
 #[inline]
 fn is_leading_surrogate(code_unit: u16) -> bool {
-    code_unit & 0xD800 == 0xD800
+    code_unit & 0xFC00 == 0xD800
```
Turns out the issue was in `is_leading_surrogate` and `is_trailing_surrogate`. Instead of checking the entire byte to match for `0xD8`, we only checked some bits. This would lead match too many patterns, for instance: `0xF8` (`0b11111000`) has all bits set of `0xD8` (`0b11011000`) but is outside of the surrogate range.

flub commented on relay-wstring/src/lib.rs:
```diff
@@ -370,6 +388,16 @@ impl fmt::Display for WStr {
+///
+/// UTF-16 can encode characters in either single units or pairs, where the first unit is a leading
+/// surrogate and the second unit is a trailing surrogate. Additionally, the entire range of
+/// `0xD800-0xDBFF` is reserved for surrogates. This means that single unit unicode code points are
```
Range should be `0xD800-0xDFFF`?

flub commented on relay-wstring/src/lib.rs:
```diff
@@ -380,10 +408,10 @@ impl fmt::Display for WStr {
 /// whether they are the *leading* or *trailing* code unit of the surrogate pair.  And for
 /// the leading surrogate this bit prefix is `110110`, thus all leading surrogates have a
-/// code unit between 0xD800-0xDBFF.
+/// code unit of `0xD8xx`.
```
I think this wording should not have been changed.  The full range here goes higher than 0xD8FF which is the highest suggested here.

flub commented on relay-wstring/src/lib.rs:
```diff
@@ -396,10 +424,10 @@ fn is_leading_surrogate(code_unit: u16) -> bool {
 /// whether tye are the *leading* or *trailing* code unit of the surrogate pair.  Anf for
 /// the trailing surrogate this bit prefix is `110111`, thus all trailing surrogates have a
-/// code unit between 0xDC00-0xDFFF.
+/// code unit of `0xDCxx`.
```
Same here, the original wording was correct and describes what the code does now.

Swatinem commented on relay-wstring/src/lib.rs:
```diff
@@ -262,18 +262,23 @@ impl<'a> Iterator for WStrChars<'a> {
         } else {
+            debug_assert!(is_leading_surrogate(u), "code unit not a leading surrogate");
+
             let chunk = self.chunks.next().expect("missing trailing surrogate");
```
we still have a potential panic here. Are we *absolutely sure* that we always get two matching surrogate pairs? AFAICR, unmatched surrogates is a quite common thing when dealing with utf-16.

Or is that case covered somewhere else?

flub replied:
Yeah, we were in two minds here.  These `.expects()` in these places just saved us by notifying us of this bug...

The reasoning to let this one is because this expects non-local code, where the other panic-avoiding code is super local and can be reasoned about locally.  Still, maybe that's the wrong call and we should avoid panicing code entirely.

Wrt the surrogate pair.  The whole premise of `WStr` is that it absolutely ensures you always have valid UTF-16.  And that means no lone surrogates.

Swatinem commented on relay-wstring/src/lib.rs:
```diff
@@ -299,18 +306,26 @@ impl<'a> DoubleEndedIterator for WStrChars<'a> {
+                "code unit not a trailing surrogate"
+            );
+
             let chunk = self.chunks.next_back().expect("missing leading surrogate");
```
same here

flub commented on relay-wstring/src/lib.rs:
```diff
@@ -370,6 +388,16 @@ impl fmt::Display for WStr {
+/// always encoded outside of this range.
+fn is_standalone_unit(code_unit: u16) -> bool {
+    code_unit < 0xD800 || code_unit > 0xDFFF
+}
```
Having reviewed all the ranges I still don't see this.  I believe all the code units for which `is_standalone_unit(u) -> false` will have either `is_leading_surrogate(u) -> true` or `is_trailing_surrogate(u) -> true`.  So I'm not sure we needed to add this.

jan-auer replied:
Yes, you're right. For I've confused `0xD800` and `0xDB00` when scanning through the comments, hence some of the confusion. You're right that we don't need this.

flub reviewed (commented):
huh, I can't request changes on my own PR?  strange github...

jan-auer committed:
fix(wstring): Fix test and address comments

jan-auer committed:
meta: Changelog
